sound/soc/sof/ipc4-pcm.c

```diff
@@ -571,11 +576,67 @@ static int sof_ipc4_pcm_get_time_info(struct snd_soc_component *component,
 	return 0;
 }
 
+static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+					    struct snd_pcm_substream *substream)
+{
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_llp_reading_slot llp, last_llp;
+	struct snd_sof_pcm_stream *stream;
+	struct snd_sof_pcm *spcm;
+	snd_pcm_uframes_t pos;
+	int time = 100;
+	u64 link_pos;
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	if (!spcm)
+		return 0;
+
+	stream = &spcm->stream[substream->stream];
+
+	if (stream->llp_offset) {
+		/*
+		 * Host dma position is gotten from mmio in host space but llp counter
+		 * is updated by fw by each schedule time, 1ms in common case. To
+		 * synchronize them, host will read host dma position when the llp counter
+		 * is just changed by fw. Driver checks the llp change every 10 us, so the
+		 * error time will about 10 ~ 15 us.
+		 */
+		sof_mailbox_read(stream->sdev, stream->llp_offset, &last_llp, sizeof(last_llp));
+		while (time--) {
+			usleep_range(10, 15);
+
+			sof_mailbox_read(stream->sdev, stream->llp_offset, &llp, sizeof(llp));
+			if (llp.reading.llp_l != last_llp.reading.llp_l)
+				break;
+		}
+	}
+
+	pos = sof_ops(stream->sdev)->pcm_pointer(stream->sdev, substream);
+
+	if (!stream->llp_offset)
+		link_pos = stream->posn.dai_posn;
+	else
+		link_pos = PACK_TWO_U32_TO_U64(llp.reading.llp_l, llp.reading.llp_u);
+
+	/*
+	 * since dma buffer is a cycle buffer, pos is wrapped but llp value
+	 * is increased linearly, need to convert pos into linear space
+	 */
+	if (pos < stream->last_pos)
+		stream->wrap_count++;
+
+	stream->last_pos = pos;
+	pos += stream->wrap_count * substream->runtime->buffer_size;
```

Comment:
Could we reuse the boundary/wrap logic in PCM core? 

I mean this patch adds a 32bit wrap. If we have a 48kHz, 4ch, 32bit stream and application programs a small period/buffer size, wrap-count itself will wrap over 2^32 after 30 days of running. Not your average case, but unless we are talking about years of runtime, I think the code needs to handle the case.

Reply:
we can do it in 64bits. The llp counter is also 64bits but I used only 32bits. The logic in PCM core is a little complex and I not sure I can exactly follow it.

Reply:
I find another simple way. We can get host dma position in hw except DPIB method